### DISCUSSION TITLE
RTEMS and Micropython

### DISCUSSION THREAD

**Discussion by rlourette** [micropython/micropython#10431] [Other]
Is anyone working on a branch/fork of Micropython for the RTEMS RTOS? Any information would be helpful. Thanks.

**Comment by rlourette** [comment]
@dpgeorge  , any thoughts on this question?

**Reply by jimmo** [reply]
@rlourette  Not that I'm aware of.

If you're interested in working on this, then feel free to ask questions here (or on the Discord).

The core team only has limited capacity to maintain the core ports in the main repo, so I would imagine that this would not be likely to be merged into the main repo (like e.g. the Zephyr port), but definitely something that could be community-maintained in a third-party repo.

The two options are to look at embedding MicroPython into an RTEMS project (see #9529), or building a from-scratch "rtems" port (like the Zephyr port).   (The distinction here is mostly which build system is the top-level).

**Reply by rlourette** [reply]
Thanks for the reply. I understand about the core team support and thank you for doing it. I am aware that a port of µPython to RTEMs [https://essr.esa.int/project/micropython-for-leon-pre-qualified-version?options=Project&search_bar=](https://essr.esa.int/project/micropython-for-leon-pre-qualified-version?options=Project&search_bar=) was created, and that it is only available to ESA member states. According to this [article](https://essr.esa.int/project/micropython-for-leon-pre-qualified-version?options=Project&search_bar=), it can be licensed by George Robotic. Here is a [paper ](https://micropython.org/resources/publish/Spacebel_MicroPythonOBCP_Dasia2018_Paper.pdf) on it.  I am an independent aerospace consultant and I would like to explore licensing it before reinventing the wheel. My first thought was to base it on the Zephyr port. 

I very much like what I see in the [embedded port](https://github.com/micropython/micropython/compare/master...dpgeorge:micropython:embed-c-port).  I worked for an aerospace firm many years ago and embedded IronPython in ground support test equipment, so I know how powerful embedded µPython would be. I am working with the Nasa Core Flight System (cFS) framework and I would ultimately want to be involved with something similar to what Spacebel did with µPython, for the cFS platform. For now, my intention is rapid prototyping of cFS applications for modeling the system as it is built out.

Is there a particular place on the MicroPython Discord  Server for the embedded µPython discussions? I just joined the server.

Thanks to you and all the rest for supporting and contributing to MicroPython.

Rich

**Reply by jimmo** [reply]
I'm sorry, I had a complete brain fail and didn't link this question to the ESA work (I'm not involved with that work and don't know much about the specifics). I've forwarded this thread to Damien.

This wounds very interesting, it would definitely be worth discussing further. Discord is a good option too.

> Is there a particular place on the MicroPython Discord Server for the embedded µPython discussions?

The `#micropython-dev` channel would be fine!

**Reply by dpgeorge** [reply]
I worked on the ESA port of MicroPython to LEON/RTEMS, and also worked with Spacebel to do the qualification.  We recently did some follow up work to get the MicroPython RTEMS port running under RTEMS 6 (the original version of RTEMS 4.8).

There are a few different aspects to the port, eg: getting MicroPython to run on LEON, making MicroPython run as an RTEMS task, adding modules to access RTEMS primitives.

I can answer more questions about it.

**Reply by rlourette** [reply]
Sorry for the delay. Was sick with Covid. I am most interested in the RTEMs port than the LEON part. 

1) Is there a fork or pull request I can look at?
2) My interest is integrating this with NASA Core Flight System (cFS) so module references to RTEMs primitives would be very useful.
3) My first effort would be to get it running on one of the platforms RTEMs works on. I just purchased a Xilinx ZYNQ Ultrascale+ [development board](https://www.avnet.com/shop/us/products/avnet-engineering-services/aes-zub-1cg-dk-g-3074457345648900673/) from Avnet. It hosts A53s and Dornerworks ported RTEMs to that platform. I would be very interested in looking at the LEON implementation on a LEON VM or whatever platform the RTEMs 6 port was done on.

Thanks for the reply @dpgeorge!
